File: typescript/test/adversarial.test.ts
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import { describe, expect, it } from "vitest";
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";
import { decode } from "../src/index.js";
import { Base85NDecodeError } from "../src/errors.js";
import { hexToBytes } from "./helpers.js";

// Adversarial decode vectors (testvectors/adversarial_vectors.json):
// multi-byte Unicode input at various positions (character-position vs.
// storage-unit discrepancies -- JS strings are UTF-16 code units, so this
// specifically exercises surrogate-pair handling), 0-length DP signals,
// invalid/reserved DP signals, and deliberately malformed escaping.

const __dirname = dirname(fileURLToPath(import.meta.url));

interface AdversarialVector {
  name: string;
  category: string;
  kind: "must_fail" | "valid";
  input_hex: string;
  error_code?: string;
  expected_hex?: string;
}

const vectorsPath = join(__dirname, "..", "..", "testvectors", "adversarial_vectors.json");
const vectors: AdversarialVector[] = JSON.parse(readFileSync(vectorsPath, "utf-8"));

// The vectors are byte-level, because the format is: a decoder's input is
// whatever arrives on the wire, and much of what arrives is not valid UTF-8.
// Each byte becomes the character of the same code unit -- the identity on
// ASCII, which is where Alphabet-N and the ignorable whitespace all live, so
// every byte from 0x80 up is one significant character outside the alphabet.
// That is the same mapping the C ABI and the Rust vector runner use.
//
// This used to be a fatal UTF-8 decoder, which quietly confined the shared
// vector set to inputs it could express. Every vector written before the
// error_precedence group is UTF-8-valid for that reason, not by choice.
function hexToByteString(hex: string): string {
  const bytes = hexToBytes(hex);
  let out = "";
  for (const b of bytes) out += String.fromCharCode(b);
  return out;
}

describe("adversarial decode vectors", () => {
  it("loads a non-trivial number of vectors", () => {
    expect(vectors.length).toBeGreaterThanOrEqual(15);
  });

  for (const v of vectors) {
    it(`${v.category}: ${v.name}`, () => {
      const input = hexToByteString(v.input_hex);

      if (v.kind === "must_fail") {
        let thrown: unknown;
        try {
          decode(input);
        } catch (e) {
          thrown = e;
        }
        expect(thrown, `expected decode() to throw for ${v.name}`).toBeInstanceOf(Base85NDecodeError);
        expect((thrown as Base85NDecodeError).code).toBe(v.error_code);
      } else {
        const expected = hexToBytes(v.expected_hex as string);
        expect(decode(input)).toEqual(expected);
      }
    });
  }
});

File: go/adversarial_test.go
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.

package base85n

import (
	"encoding/json"
	"errors"
	"os"
	"path/filepath"
	"testing"
)

// Adversarial decode vectors (testvectors/adversarial_vectors.json):
// multi-byte Unicode input at various positions (character-position vs.
// storage-unit discrepancies), 0-length DP signals, invalid/reserved DP
// signals, and deliberately malformed escaping.

type adversarialVector struct {
	Name        string `json:"name"`
	Category    string `json:"category"`
	Kind        string `json:"kind"` // "must_fail" or "valid"
	InputHex    string `json:"input_hex"`
	ErrorCode   string `json:"error_code"`
	ExpectedHex string `json:"expected_hex"`
}

func loadAdversarialVectors(t *testing.T) []adversarialVector {
	t.Helper()
	path := filepath.Join("..", "testvectors", "adversarial_vectors.json")
	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("failed to read adversarial vectors at %s: %v", path, err)
	}
	var vecs []adversarialVector
	if err := json.Unmarshal(data, &vecs); err != nil {
		t.Fatalf("failed to parse adversarial vectors: %v", err)
	}
	if len(vecs) < 15 {
		t.Fatalf("expected a non-trivial adversarial vector set, got %d", len(vecs))
	}
	return vecs
}

// sentinelForErrorCode maps the shared error_code string (identical across
// every language implementation) to this package's sentinel error.
func sentinelForErrorCode(code string) error {
	switch code {
	case "invalid_character":
		return ErrInvalidCharacter
	case "unexpected_end_of_stream":
		return ErrUnexpectedEOF
	case "undefined_signal":
		return ErrUndefinedSignal
	case "invalid_final_block":
		return ErrInvalidFinalBlock
	default:
		return nil
	}
}

func TestAdversarialVectors(t *testing.T) {
	vectors := loadAdversarialVectors(t)

	for _, v := range vectors {
		v := v
		t.Run(v.Name, func(t *testing.T) {
			inputBytes := mustHexDecode(t, v.InputHex)
			input := string(inputBytes) // Go strings are just byte sequences; preserves the UTF-8 bytes as-is.

			switch v.Kind {
			case "must_fail":
				want := sentinelForErrorCode(v.ErrorCode)
				if want == nil {
					t.Fatalf("unknown error_code %q", v.ErrorCode)
				}
				got, err := Decode(input)
				if err == nil {
					t.Fatalf("expected Decode to fail with %v, but it succeeded with %v", want, got)
				}
				if !errors.Is(err, want) {
					t.Fatalf("expected error category %v, got %v", want, err)
				}
			case "valid":
				expected := mustHexDecode(t, v.ExpectedHex)
				got, err := Decode(input)
				if err != nil {
					t.Fatalf("expected Decode to succeed, got error %v", err)
				}
				if string(got) != string(expected) {
					t.Fatalf("decoded bytes did not match expected_hex: got %x, want %x", got, expected)
				}
			default:
				t.Fatalf("unknown vector kind %q", v.Kind)
			}
		})
	}
}
